### ISSUE TITLE
"Some of Tor Browser's security features may offer less protection on your current operating system."

### ISSUE THREAD

**Issue by Midar** [flathub/org.torproject.torbrowser-launcher#67]
This happens since the update to Tor Browser 14. It comes as a popup on every start. It has a "How to fix this issue" link which points to Mozilla's instructions on how to configure AppArmor. However, I'm not using AppArmor, so I guess the problem is Flatpak instead.

**Comment by AsciiWolf** [comment]
Yeah, I just got it too. AppArmor is just one of the things mentioned in the Mozilla article so I am not sure it is related. And considering that AppArmor is basically a Debian-systems specific thing, it would be quite stupid in my opinion.

Anyway, this is most likely some upstream Firefox thing that was not intended to be a part of Tor Browser and will hopefully be removed.

/cc @boklm

**Comment by boklm** [comment]
> This happens since the update to Tor Browser 14. It comes as a popup on every start. It has a "How to fix this issue" link which points to Mozilla's instructions on how to configure AppArmor. However, I'm not using AppArmor, so I guess the problem is Flatpak instead.

Which Linux distribution are you using?

It looks like Ubuntu 24.04 disabled the use of user namespaces, unless
allowed by some AppArmor profile. So if you're running Ubuntu 24.04 this
might be unrelated to using Flatpak.

If you're running an other OS that doesn't disable user namespaces by
default, this might be related to using Flatpak.

We had an other report of a similar issue:
https://gitlab.torproject.org/tpo/applications/tor-browser/-/issues/43235

**Comment by AsciiWolf** [comment]
I am using Fedora 40.

**Comment by boklm** [comment]
It looks like user namespaces doesn't work inside flatpaks:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1756236
* https://bugzilla.mozilla.org/show_bug.cgi?id=1882881
* https://bugzilla.mozilla.org/show_bug.cgi?id=1909616

**Comment by Midar** [comment]
Fedora 41 in my case.

Interestingly enough, no Firefox version ever showed this warning for me, not even 128 before it was ESR.

It seems this is the change? https://hg.mozilla.org/integration/autoland/rev/fcd48a1ad31e

**Comment by CocolinoFan** [comment]
Same issue on Kubuntu oracular 24.10 x86_64 - Linux 6.11.0-9-generic
AppArmor is really cool and makes me feel safe. Tor should work with AppArmor.

**Comment by AsciiWolf** [comment]
@boklm Any chance this warning message could be disabled for TBB running inside a Flatpak container (just like we fixed the auto update feature)?

**Comment by rusty-snake** [comment]
`security.sandbox.warn_unprivileged_namespaces = false`

**Comment by anonym** [comment]
> `security.sandbox.warn_unprivileged_namespaces = false`

tl;dr: I think this is an ok solution for now.

That will indeed _hide_ the issue, but the issue is real. If you go to `about:support#sandbox` you'll see that "User Namespaces" is false, so the sandbox is degraded. The reason is that Flatpak's seccomp filter blocks the browser processes from using unprivileged user namespaces.

Some relevant Firefox tickets:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1756236
* https://bugzilla.mozilla.org/show_bug.cgi?id=1882881

So Mozilla is working on a solution, with a major part of it being the Fork()Server ([nice FOSDEM presentation](https://www.youtube.com/watch?v=s7Pn6OpR-0c)) which [supposedly will be ready soon](https://www.phoronix.com/news/Firefox-ForkServer-Linux-Nears). So it might be a while still until it is fixed in Firefox, and yet another while until it hits an ESR release that Tor Browser is based on. So this won't go away on its own any time soon.

I am not 100% sure of the following analysis but hear goes: while running Tor Browser as a flatpak, so without the user namespace part of the browser sandbox, essentially what you lose is isolation between the processes (but the seccomp filter is still working). So instead of each process being isolated from each other (and the rest of the system) by running in their own namespaces they now share the same namespace created by flatpak, which still is isolated against the rest of the system. So it clearly is a degraded sandbox, but maybe not as bad as the warning banner implies, which I suspect was designed for situations where the browser processes wouldn't be isolated _at all_ (neither against each other nor the rest of the system), i.e. if you (or your distro) has disabled unprivileged user namespaces or if you run an ancient kernel without support.

Note that the official `org.mozilla.firefox` flatpak does _not_ show this warning, which IMHO is a decent enough reason for Tor Browser to hide it. (They don't set this pref, so they are doing it in some other way that I haven't bother looking into, but we know they too are affected by what is reported in `about:support#sandbox`.)
